chain-tx-enclave-next/mls/src/tree.rs

```diff
@@ -335,34 +441,264 @@ impl Tree {
+                self.cs
+                    .expand_label(
+                        &leaf_private_key.marshal(),
+                        vec![],
+                        "path",
+                        &[],
+                        self.cs.secret_size(),
+                    )
+                    .expect("invalid length"),
+            )
+        } else {
+            let ancestor = NodeSize::ancestor_of(sender, self.my_pos);
+            let my_node = if self.my_pos > sender {
+                ancestor.right(leaf_len).expect("ancestor not leaf node")
+            } else {
+                ancestor.left().expect("ancestor not leaf node")
+            };
+            let path_node = NodeSize::from_leaf_index(sender)
+                .direct_path(leaf_len)
+                .into_iter()
+                .zip(path_nodes.iter())
+                .find(|(n, _)| *n == ancestor)
+                .expect("invalid path nodes")
+                .1;
+            for (i, node_index) in resolve(&self.nodes, my_node).iter().copied().enumerate() {
+                if let Some(private_key) = self.node_private_key(node_index, leaf_private_key) {
+                    let secret = self
+                        .cs
+                        .decrypt(private_key, ctx, &path_node.encrypted_path_secret[i])
+                        .expect("decrypt failed");
+                    return (ancestor, SecretVec::new(secret));
+                }
+            }
+            panic!("don't find private key to decrypt secret");
+        }
+    }
+
+    /// returns (leaf_parent_hash, commit_secret)
+    pub fn apply_path_secrets(
+        &mut self,
+        sender: LeafSize,
+        ctx: &[u8],
+        path_nodes: &[DirectPathNode],
+        leaf_private_key: &HPKEPrivateKey,
+    ) -> (Vec<u8>, SecretVec<u8>) {
+        let leaf_len = self.leaf_len();
+        let from = NodeSize::from_leaf_index(sender);
+        // set public key
+        for (node_index, path_node) in from
+            .direct_path(leaf_len)
+            .into_iter()
+            .zip(path_nodes.iter())
+        {
+            self.nodes[node_index.0].merge_public(path_node.public_key.clone());
+        }
+
+        // Find overlap and decrypt the path secret
+        let (overlap, mut last_secret) =
+            self.decrypt_path_secrets(sender, ctx, path_nodes, leaf_private_key);
+
+        // implant the path secret
+        self.nodes[overlap.0]
+            .parent_node_mut()
+            .expect("implant invalid node")
+            .set_secret(self.cs, &last_secret);
+        for node_index in overlap.direct_path(leaf_len).iter() {
+            last_secret = self
+                .cs
+                .expand_label(&last_secret, vec![], "path", &[], self.cs.secret_size())
+                .expect("invalid length");
+            self.nodes[node_index.0]
+                .parent_node_mut()
+                .expect("implant invalid node")
+                .set_secret(self.cs, &last_secret);
+        }
```

Comment:
A bit confused about this part -- my assumption is, as this is called by the recipient in processing commit in group init, that this is "Synchronizing Views of the Tree".
There's one step:
> The recipient SHOULD verify that the received public keys agree with the public keys derived from the new path_secret values.

I'm more sure whether that's done?

Reply:
Changed the `set_secret` to `set_secret_verify_public`, the verification is done inside. the public key is updated in `merge_public_key`.